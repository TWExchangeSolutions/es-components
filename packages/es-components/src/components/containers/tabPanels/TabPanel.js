import React from 'react';
import styled from 'styled-components';
import Tab from './Tab';

const TabWrapper = styled.div`
  display: flex;
`;

const TabFormatter = styled.div`
  display: flex;
  flex-direction: column;
  font-size: inherit;
  overflow: hidden;
  @media (max-width: ${props => props.theme.screenSize.desktop}) {
    width: 100%;
  }
  @media (min-width: ${props => props.theme.screenSize.desktop}) {
    flex-direction: row;
    justify-content: flex-start;
  }
`;

const TabContent = styled.div`
  margin-top: -1px;
  background-color: ${props => props.theme.colors.white};
  overflow: auto;
  border-top: 1px solid ${props => props.theme.colors.gray4};
`;

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    const child = Array.isArray(this.props.children)
      ? props.children[0]
      : props.children;
    this.state = {
      value: child.props.name,
      currentContent: child.props.children
    };
    this.tabChanged = this.tabChanged.bind(this);
  }

  tabChanged(name, child) {
    this.setState({
      value: name,
      currentContent: child
    });
  }

  render() {
    const { children } = this.props;
    const elements = React.Children.map(children, (child, i) => {
      const isSelected = child.props.name === this.state.value;
      return React.cloneElement(child, {
        key: child.props.name,
        selected: isSelected,
        action: this.tabChanged,
        selectedName: this.state.value
      });
    });

    return (
      <div className="es-tab-panel">
        <TabWrapper className="es-tab-panel__wrapper">
          <TabFormatter className="es-tab-panel__tabs">{elements}</TabFormatter>
        </TabWrapper>
        <TabContent className="es-tab-panel__content">
          {this.state.currentContent}
        </TabContent>
      </div>
    );
  }
}

function childrenRule(props, propName, component) {
  let children = props[propName];
  if (!Array.isArray(children)) {
    children = [children];
  }

  if (
    !children.every(
      child => child.type.displayName === 'Tab' || child.type.target === Tab
    )
  ) {
    return new Error('Tab Panel only accepts Tabs as direct descendants.');
  }

  return null;
}

TabPanel.propTypes = {
  /**
   * Makes sure immediate children are Tab or Tab List, as we cannot render anything else in the tab heading.
   */
  children: childrenRule
};

TabPanel.Tab = Tab;

export default TabPanel;
