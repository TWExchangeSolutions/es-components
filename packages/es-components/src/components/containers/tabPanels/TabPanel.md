Tab Panels allow us to store lots more data on the screen and only show the amount we want, when we want. This works by having a TabPanel.Tab which provides a quick peek into what information will be display in the content area when clicked. The TabPanel.Tab has a name, which is a string or JSX to be displayed, and an optional simpleName, which is a string that will be read by a screen reader. The TabPanel.Tab can have children inside of it that will be rendered inside of the TabPanel's content area.

```
<TabPanel>
  <TabPanel.Tab name="Hi there">
    <p>
      HELLO WORLD!!!!
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab simpleName="interesting" name={<div><Icon name="certificate" />Interesting 5 plans</div>}>
    <div>
      <p>
        Nesting is fun
      </p>
    </div>
  </TabPanel.Tab>
  <TabPanel.Tab name="Lorem ipsum" announcerText="Tab is now selected">
    <p>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum
    </p>
  </TabPanel.Tab>
  <TabPanel.Tab name="dolor sit">
    <p>
      Multiple paragraphs inside the Tab.
    </p>
    <p>
      ed ut perspiciatis unde omnis iste natus error sit voluptatem accusantium doloremque laudantium, totam rem aperiam, eaque ipsa quae ab illo inventore veritatis et quasi architecto beatae vitae dicta sunt explicabo. Nemo enim ipsam voluptatem quia voluptas sit aspernatur aut odit aut fugit, sed quia consequuntur magni dolores eos qui ratione voluptatem sequi nesciunt. Neque porro quisquam est, qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit, sed quia non numquam eius modi tempora incidunt ut labore et dolore magnam aliquam quaerat voluptatem. Ut enim ad minima veniam, quis nostrum exercitationem ullam corporis suscipit laboriosam, nisi ut aliquid ex ea commodi consequatur? Quis autem vel eum iure reprehenderit qui in ea voluptate velit esse quam nihil molestiae consequatur, vel illum qui dolorem eum fugiat quo voluptas nulla pariatur?
    </p>
  </TabPanel.Tab>
</TabPanel>
```