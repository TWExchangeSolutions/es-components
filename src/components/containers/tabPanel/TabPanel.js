import React from 'react';
import PropTypes from 'prop-types';
import { colors } from '../../theme';
import styled from 'styled-components';
import Tab from './Tab';
import TabList from './TabList';

const ContentDiv = styled.div`
  border-top: 1px solid ${colors.grayLight};
`;

function TabItem({ name, children }) {
  return <span />;
}

TabItem.propTypes = {
  name: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element)
  ])
};

class TabPanel extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      value: '',
      allValues: []
    };

    this.tabChanged = this.tabChanged.bind(this);
  }

  componentDidMount() {
    const values = [];
    React.Children.forEach(this.props.children, (child, i) => {
      if (child.type === Tab) {
        values.push(child.props.name);
      }
      if (child.type === TabList) {
        React.Children.forEach(child.props.children, (secondChild, j) => {
          values.push(secondChild.props.name);
        });
      }
      this.setState({ allValues: values });
    });
  }

  setupContent(child) {
    return (
      <ContentDiv id="tab-panel-content" aria-live="polite">
        {child}
      </ContentDiv>
    );
  }

  tabChanged = name => {
    this.setState({ value: name });
  };

  isChildSelected(value, index) {
    return value === this.state.value || (!this.state.value && index === 0);
  }

  handleKeyDown = event => {
    let newIndex = 0;
    const getIndexFromMoreBrowsers = offset => {
      this.state.allValues.some((val, i) => {
        if (val === this.state.value) {
          newIndex = i + offset;
        }
        return false;
      });
    };
    if (event.key === 'ArrowLeft') {
      event.preventDefault();
      getIndexFromMoreBrowsers(-1);
      const finalIndex =
        newIndex < 0 ? this.state.allValues.length - 1 : newIndex;
      this.setState({ value: this.state.allValues[finalIndex] });
    } else if (event.key === 'ArrowRight') {
      event.preventDefault();
      getIndexFromMoreBrowsers(1);
      const finalIndex = newIndex >= this.state.allValues.length ? 0 : newIndex;
      this.setState({ value: this.state.allValues[finalIndex] });
    }
  };

  render() {
    let content = null;
    const elements = React.Children.map(this.props.children, (child, i) => {
      let selected = false;
      if (child.type === Tab) {
        if (this.isChildSelected(child.props.name, i)) {
          content = this.setupContent(child.props.children);
          selected = true;
        }
        return React.cloneElement(child, {
          selected,
          action: this.tabChanged,
          key: child.props.name
        });
      }
      if (child.type === TabList) {
        const options = React.Children.map(
          child.props.children,
          (secondChild, j) => {
            if (this.isChildSelected(secondChild.props.name, i)) {
              content = this.setupContent(secondChild.props.children);
              selected = true;
            }
            return { optionText: secondChild.props.name };
          }
        );
        return React.cloneElement(child, {
          selected,
          action: this.tabChanged,
          options,
          key: child.props.name
        });
      }
      return null;
    });

    return (
      <div onKeyDown={this.handleKeyDown} role="application">
        {elements}
        {content}
      </div>
    );
  }
}

function childrenRule(props, propName, component) {
  if (
    !props[propName].every(
      child => child.type.name === 'Tab' || child.type.name === 'TabList'
    )
  ) {
    return new Error(
      'Tab Panel only accepts Tab or TabList as direct descendants.'
    );
  }
  return null;
}

TabPanel.propTypes = {
  children: childrenRule
};

TabPanel.Tab = Tab;
TabPanel.TabList = TabList;
TabPanel.TabItem = TabItem;

export default TabPanel;
