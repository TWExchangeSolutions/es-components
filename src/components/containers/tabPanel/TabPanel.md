
Testing text

```

<TabPanel>
  <TabPanel.Tab name="Home">
    Lorem uipsyunkjaflklsadfkahs
  </TabPanel.Tab>
  <TabPanel.Tab name="Menu">
    <div>
      Food truck fixie locavore, accusamus mcsweeney's marfa nulla single-origin coffee squid. Exercitation +1 labore 
    </div>
  </TabPanel.Tab>
  <TabPanel.Tab name="Display">
    <div>
      <p>
        Etsy mixtape wayfarers, ethical wes anderson tofu before they sold out mcsweeney's organic lomo retro fanny pack
      </p>
      <ul>
        <li>Item first</li>
        <li>ItemSecond</li>
      </ul>
    </div>
  </TabPanel.Tab>
  <TabPanel.TabList name="many options">
    <TabPanel.TabItem name="option1">
      <div>
        Hello there.
      </div>
    </TabPanel.TabItem>
    <TabPanel.TabItem name="option2">
      <div>
        Etsy mixtape wayfarers, ethical wes anderson tofu before they sold out mcsweeney's organic lomo retro fanny pack lo-fi farm-to-table readymade. Messenger bag gentrify pitchfork tattooed craft beer, iphone skateboard locavore carles etsy salvia banksy hoodie helvetica. DIY synth PBR banksy irony. <a href ="#hi">extra link</a> Leggings gentrify squid 8-bit cred pitchfork. Williamsburg banh mi whatever gluten-free, carles pitchfork biodiesel fixie etsy retro mlkshk vice blog. Scenester cred you probably haven't heard of them, vinyl craft beer blog stumptown. Pitchfork sustainable tofu synth chambray yr
      </div>
    </TabPanel.TabItem>
    <TabPanel.TabItem name="option3">
      <div>
        Humbug
      </div>
    </TabPanel.TabItem>
  </TabPanel.TabList>
</TabPanel>
```