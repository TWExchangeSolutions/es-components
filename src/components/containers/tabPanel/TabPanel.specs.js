/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import TabPanel from './TabPanel';

const buildTabPanel = props => (
  <TabPanel>
    <TabPanel.Tab name="tab1">first</TabPanel.Tab>
    <TabPanel.Tab name="tab2" className="testing">
      second
    </TabPanel.Tab>
    <TabPanel.Tab name="tab3">third</TabPanel.Tab>
    <TabPanel.TabList name="multiTab">
      <TabPanel.TabItem name="option1" value="TheFirst">
        <div>multi first</div>
      </TabPanel.TabItem>
      <TabPanel.TabItem name="option2" value="TheSecond">
        <div>multi second</div>
      </TabPanel.TabItem>
      <TabPanel.TabItem name="option3" value="TheThird">
        <div>multi third</div>
      </TabPanel.TabItem>
    </TabPanel.TabList>
  </TabPanel>
);

describe('Tab panel', () => {
  let panelInstance;

  beforeAll(() => {
    panelInstance = mount(buildTabPanel());
  });

  it('renders as expected', () => {
    const tree = renderer.create(buildTabPanel()).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('clicking a tab should set the panels value to its name', () => {
    const secondTab = panelInstance.find('.testing');
    expect(panelInstance.state('value')).toBe('');
    secondTab.simulate('click');
    expect(panelInstance.state('value')).toBe('tab2');
  });

  it('child is selected correctly', () => {
    panelInstance.setState({ value: 'something' });
    expect(panelInstance.instance().isChildSelected('something', 5)).toBe(true);
  });

  it('child is not selected when values are different', () => {
    panelInstance.setState({ value: 'test' });
    expect(panelInstance.instance().isChildSelected('something', 5)).toBe(
      false
    );
  });

  it('child is selected when default', () => {
    panelInstance.setState({ value: '' });
    expect(panelInstance.instance().isChildSelected('something', 0)).toBe(true);
  });

  it('decrements when left arrow is pressed', () => {
    panelInstance.setState({ value: 'bye', allValues: ['hi', 'bye', 'there'] });
    panelInstance.simulate('keyDown', { key: 'ArrowLeft' });
    expect(panelInstance.state('value')).toBe('hi');
  });

  it('increments when right arrow is pressed', () => {
    panelInstance.setState({ value: 'bye', allValues: ['hi', 'bye', 'there'] });
    panelInstance.simulate('keyDown', { key: 'ArrowRight' });
    expect(panelInstance.state('value')).toBe('there');
  });

  it('decrements and wraps when left arrow is pressed at start', () => {
    panelInstance.setState({ value: 'hi', allValues: ['hi', 'bye', 'there'] });
    panelInstance.simulate('keyDown', { key: 'ArrowLeft' });
    expect(panelInstance.state('value')).toBe('there');
  });

  it('increments and wraps when right arrow is pressed at end', () => {
    panelInstance.setState({
      value: 'there',
      allValues: ['hi', 'bye', 'there']
    });
    panelInstance.simulate('keyDown', { key: 'ArrowRight' });
    expect(panelInstance.state('value')).toBe('hi');
  });
});
