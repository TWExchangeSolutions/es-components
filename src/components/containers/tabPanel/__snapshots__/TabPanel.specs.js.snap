// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab panel renders as expected 1`] = `
<div
  onKeyDown={[Function]}
  role="application"
>
  <button
    aria-controls="tab-panel-content"
    className="sc-bdVaJa fVOfLO"
    onClick={[Function]}
    selected={true}
  >
    tab1
  </button>
  <button
    aria-controls="tab-panel-content"
    className="testing sc-bdVaJa dhbpgv"
    onClick={[Function]}
    selected={false}
  >
    tab2
  </button>
  <button
    aria-controls="tab-panel-content"
    className="sc-bdVaJa dhbpgv"
    onClick={[Function]}
    selected={false}
  >
    tab3
  </button>
  <select
    className="sc-bwzfXH gElLmT"
    onChange={[Function]}
    selected={false}
    value="multiTab"
  >
    <option
      disabled={true}
    >
      multiTab
    </option>
    <option
      value="option1"
    >
      option1
    </option>
    <option
      value="option2"
    >
      option2
    </option>
    <option
      value="option3"
    >
      option3
    </option>
  </select>
  <div
    aria-live="polite"
    className="sc-htpNat iTEfCw"
    id="tab-panel-content"
  >
    first
  </div>
</div>
`;
